pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,350 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package basicstationlns
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/gorilla/websocket"
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/basicstation"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io/basicstationlns/messages"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+	"go.thethings.network/lorawan-stack/pkg/web"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	errEmptyGatewayEUI           = errors.Define("empty_gateway_eui", "empty gateway EUI")
+	errMessageTypeNotImplemented = errors.DefineUnimplemented("message_type_not_implemented", "message of type `{type}` is not implemented")
+)
+
+type srv struct {
+	ctx      context.Context
+	server   io.Server
+	upgrader *websocket.Upgrader
+	tokens   io.DownlinkTokens
+}
+
+func (*srv) Protocol() string { return "basicstation" }
+
+// New returns a new Basic Station frontend that can be registered in the web server.
+func New(ctx context.Context, server io.Server) web.Registerer {
+	ctx = log.NewContextWithField(ctx, "namespace", "gatewayserver/io/basicstation")
+	return &srv{
+		ctx:      ctx,
+		server:   server,
+		upgrader: &websocket.Upgrader{},
+	}
+}
+
+func (s *srv) RegisterRoutes(server *web.Server) {
+	group := server.Group(ttnpb.HTTPAPIPrefix + "/gs/io/basicstation")
+	group.GET("/discover", s.handleDiscover)
+	group.GET("/traffic/:id", s.handleTraffic)
+}
+
+func (s *srv) handleDiscover(c echo.Context) error {
+	ctx := c.Request().Context()
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "discover",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
+		return err
+	}
+	defer ws.Close()
+
+	_, data, err := ws.ReadMessage()
+	if err != nil {
+		logger.WithError(err).Debug("Failed to read message")
+		return err
+	}
+	var req messages.DiscoverQuery
+	if err := json.Unmarshal(data, &req); err != nil {
+		logger.WithError(err).Debug("Failed to parse discover query message")
+		return err
+	}
+
+	if req.EUI.IsZero() {
+		writeDiscoverError(s.ctx, ws, "Invalid request")
+		return errEmptyGatewayEUI
+	}
+
+	ids := ttnpb.GatewayIdentifiers{
+		EUI: &req.EUI.EUI64,
+	}
+	ctx, ids, err = s.server.FillGatewayContext(ctx, ids)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to fill gateway context")
+		writeDiscoverError(ctx, ws, "Router not provisioned")
```

Comment:
Again, is this because of spec? Otherwise include the err.

Reply:
Here also, the actual error message is not processed but just logged. 
The current way of handling it in the station is to retry for a certain no of times and then shutdown.